### PR TITLE
Generate loop start, end locations in llvm.loop metadata, using new metadata

### PR DESCRIPTION
**Pull request by michalpasztamobica**
This PR is a continuation of https://github.com/flang-compiler/flang/pull/976
I rebased that PR on top of recent metadata changes from @Chirag-Khandelwal and adjusted it to use new metadata.
I humbly ask for a detailed review as this is the first time I touched the llvm IR generation.
In particular, please pay attention to the adjustments in `ivdep.f90` and `omp_simd.f90` - These tests were failing already in #976 . To my understanding the changes in the code generation were intentional and the tests needed to be adjusted, but I am not sure if I am right.

### CODE REVIEW

**Review by pawosm-arm — APPROVED** [review]
It looks fine, although "Replace llvm's mem.parallel_loop_access with group.access" commit message does not say why it is needed to do so.

**Review by pawosm-arm — CHANGES_REQUESTED** [review]
it fails to build. Incomplete commit?

**Comment by michalpasztamobica** [comment]
I somehow lost one line during rebase... But now it's back in - everything builds and passes tests :-) 
@pawosm-arm I also amended the commit message giving a link to the changes in llvm which triggered this.

**Review by pawosm-arm — APPROVED** [review]

**Comment by michalpasztamobica** [comment]
@Chirag-Khandelwal , I cannot add you to the reviewers list, but your confirmation that this makes sense will be very valuable :)

**Comment by Chirag-Khandelwal** [comment]
It looks great to me! :) @michalpasztamobica

**Comment by michalpasztamobica** [comment]
@bryanpkc , @SouraVX , @peterwaller-arm , @kiranchandramohan , your feedback on this will be much appreciated. Thank you in advance.

**Comment by liuxiao0909c** [comment]
Hi, @michalpasztamobica , After the commit https://github.com/flang-compiler/flang/pull/1094/commits/efa0d02e4a44b3896260b0287bcc212a55bdce69 ,  both `write_no_depcheck_metadata` and `write_llaccgroup_metadata`(added in commit https://github.com/flang-compiler/flang/commit/f434474f1f071be10e82689c5249408f234517b5 ) will generate `!llvm.access.group` metadata,  is there any possibility of a merger?

**Comment by michalpasztamobica** [comment]
Hi @liuxiao0909c . Thanks for the review. 
Indeed, the functions were similar - I removed the older `write_no_depcheck_metadata` and parametrized `write_llaccgroup_metadata` to be more flexible. I hope this is OK now?

**Comment by michalpasztamobica** [comment]
In yesterday's call @bryanpkc rightly pointed out that the new metadata might be breaking compatibility with older llvm versions.
However - I noticed that `llvm.access.group` is already [available in llvm-10](https://releases.llvm.org/10.0.0/docs/LangRef.html#llvm-access-group-metadata), so perhaps the situation is not that bad? @bryanpkc , would you be able to fully assess the risk for llvm-10 and let us know if this path is OK for you? I believe only Huawei is currently interested in llvm-10 support?

I don't think it makes sense to introduce release branches like we do in llvm-project-classic-flang, but if necessary we could add a CMake flag, to address code differences for different llvm versions?

File: tools/flang2/flang2exe/cgmain.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/cgmain.cpp:3341]: Please use spaces instead of tabs for indentation, and left-align multi-line parameter lists.

File: tools/flang2/flang2exe/cgmain.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/cgmain.cpp:2924]: Please use spaces instead of tabs for indentation, and left-align multi-line parameter lists.

Comment (michalpasztamobica) [reply tools/flang2/flang2exe/cgmain.cpp]: Thank you for catching this! Fixed now.

File: test/directives/unroll.f90
Comment (bryanpkc) [inline test/directives/unroll.f90:78]: Can these lines use the FileCheck directive `CHECK-O0-DAG`? In my downstream build, the test outputs these lines in a different order:
```
!15 = !{!15, !12, !14, !16}
!16 = !{!"llvm.loop.unroll.enable"}
```

File: test/f90_correct/bih-head.f90
Comment (kiranchandramohan) [inline test/f90_correct/bih-head.f90:1]: Can we move this test to a new directory (test/ili_correct) since this tests the ILI output? I am OK if it is done in a subsequent patch.

**Review by kiranchandramohan — COMMENTED** [review]
@michalpasztamobica did you mention that there is a loop that does not vectorize after these changes? Could you point me to that test?

File: test/f90_correct/bih-head.f90
Comment (bryanpkc) [reply test/f90_correct/bih-head.f90:1]: +1. I was going to write the exact same comment.

**Comment by bryanpkc** [comment]
> However - I noticed that llvm.access.group is already available in llvm-10, so perhaps the situation is not that bad? @bryanpkc , would you be able to fully assess the risk for llvm-10 and let us know if this path is OK for you?

@michalpasztamobica Thanks for linking the LangRef for LLVM 10. I have merged these patches into my downstream LLVM 10-based build, but the test cases `ivdep.f90`, `omp_simd.f90`, and `vector_directive3.f90` are broken:
```
.../llvm/lib/Analysis/LoopInfo.cpp:576: llvm::Loop::isAnnotatedParallel() const::<lambda(llvm::MDNode*)>: Assertion `isValidAsAccessGroup(AccGroup) && "List item must be an access group"' failed.
```
I will re-try a vanilla LLVM 10 build and see if it behaves the same way.

File: test/directives/unroll.f90
Comment (michalpasztamobica) [reply test/directives/unroll.f90]: Done, I hope I got it right? :)

**Comment by michalpasztamobica** [comment]
@kiranchandramohan , I moved the test and squashed that change into Peter's original commit.
Regarding the vectorization, in `ivdep.f90` and `omp_simd.f90` I had to remove the `VF` regex to get the tests passing, like so:
```
! CHECK:  {{.*}} add nsw <[[VF:[0-9]+]] x i32>{{.*}}
```
to
```
! CHECK:  {{.*}} add nsw i32{{.*}}
```

File: test/directives/unroll.f90
Comment (bryanpkc) [reply test/directives/unroll.f90]: Yes, thanks.

**Comment by kiranchandramohan** [comment]
Do you get a warning like the one below?

```
DILocation not allowed within this metadata node
!30 = distinct !{!30, !22, !23, !16, !31}
!16 = !DILocation(line: 10, column: 1, scope: !5)
warning: ignoring invalid debug info in /tmp/ivdep-383a70.ll
```

**Comment by bryanpkc** [comment]
Nit: Would you mind rewording the commit currently titled "Adjust Chirag's changes to Peter's"? Please summarize what is changed in the commit title, and credit original authors in the message body.

**Comment by kiranchandramohan** [comment]
For llvm.loop and llvm.access.group, the metadata generated seems to be the same (!33). That does not look OK to me. Is this expected?
```
%53 = load i32, i32* %.dY0001_313, align 4, !llvm.access.group !33, !tbaa !42, !dbg !34
store i32 %54, i32* %.dY0001_313, align 4, !llvm.access.group !33, !tbaa !42, !dbg !34
br i1 %55, label %L.LB1_311, label %L.LB1_312, !llvm.loop !33, !dbg !34
```

**Comment by michalpasztamobica** [comment]
@bryanpkc  - sure, I will reword it, thanks for pointing this out.

@kiranchandramohan , the warnings do not show up when running `make check-all`, but I could now see them when manually compiling `ivdep.f90`. I tried to figure out which change could be an issue, but I couldn't make it out from the code. 
I then tried the original patches from https://github.com/kiranchandramohan/flang/tree/loop-line and got the same results: warning from the compiler and same value for `llvm.access.group` and `llvm.loop`. @peterwaller-arm , can you help us here?

**Comment by kiranchandramohan** [comment]
The warnings are probably coming because the llvm.access.group contains location metadata which it is not permitted to contain. 

We have to first find out why the llvm.access.group metadata is pointing to the same metadata as the llvm.loop, that should not happen. Since llvm.access.group is being introduced in the latest patch for ivdep, we have to correct the latest patch. It is probably an incorrect interaction between the code that we added and the code that chirag added.

**Comment by kiranchandramohan** [comment]
We cannot directly replace llvm.mem.parallel_loop_access with llvm.access.group as implemented in https://github.com/flang-compiler/flang/pull/1094/commits/26dc6c7da5bfb3ab2584504219f5fefa4ec6404e

The parameter of llvm.mem.parallel_loop_access was a loop id. But the parameter of llvm.access.group is a group id that is independent of the loop. The loop then should have the llvm.loop.parallel_accesses metadata which refers to the access groups. I see that the parallel_accesses metadata is generated but it is not referring to the access groups.

What we have now is something like the following for ivdep.f90,
```
        store i32 %76, i32* %.dY0001_313, align 4, !llvm.access.group !35, !dbg !34
        %77 = load i32, i32* %.dY0001_313, align 4, !llvm.access.group !35, !dbg !34
        ...
        br i1 %78, label %L.LB1_311, label %L.LB1_312, !llvm.loop !35, !dbg !34
        
!27 = distinct !{  }
!28 = !DILocation(line: 10, column: 1, scope: !25)
!30 = !{ !"llvm.loop.vectorize.enable", i1 1 }
!31 = !{ !"llvm.loop.parallel_accesses", !27 }
!34 = !DILocation(line: 12, column: 1, scope: !25)
!35 = !{ !35, !30, !31, !28, !34 }
``` 

What we should ideally have is the following,
```
        store i32 %76, i32* %.dY0001_313, align 4, !llvm.access.group !27, !dbg !34 <----- Note the change here
        %77 = load i32, i32* %.dY0001_313, align 4, !llvm.access.group !27, !dbg !34 <----- and here. Refers to a group id not loop id now.
        ...
        br i1 %78, label %L.LB1_311, label %L.LB1_312, !llvm.loop !35, !dbg !34

!27 = distinct !{  }        
!28 = !DILocation(line: 10, column: 1, scope: !25)
!30 = !{ !"llvm.loop.vectorize.enable", i1 1 }
!31 = !{ !"llvm.loop.parallel_accesses", !27 } 
!34 = !DILocation(line: 12, column: 1, scope: !25)
!35 = !{ !35, !28, !34, !30, !31 } <----- Note the re-ordering here. Location nodes should come before property nodes.
```

File: tools/flang2/flang2exe/cgmain.cpp
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/cgmain.cpp:1339]: Nit: This comment has to be updated. We do not generate llvm.mem.parallel_loop_access now.

File: tools/flang2/flang2exe/cgmain.cpp
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/cgmain.cpp:2926]: insn->flags seems to be not set when flag_check is LDST_HAS_ACCESSGRP_METADATA.

This ultimately happens because _mark_rw_access_grp_ is only called when 191, 0x4 is set (only if the directive is _vector always_). Could you call _mark_rw_access_grp_ for all of the cases where the access group metadata has to be generated??

```
    if (XBIT(191, 0x4)) {
      fix_nodepchk_flag(bih);
      mark_rw_access_grp(bih);
    }
```

**Review by kiranchandramohan — CHANGES_REQUESTED** [review]

**Comment by michalpasztamobica** [comment]
I have just pushed changes suggested off-line by @kiranchandramohan (I would have never found this fix. Thanks, Kiran!) I squashed the changes to the "Replace llvm's mem.parallel_loop_access with group.access" commit and mentioned the authorship in the commit's message. I hope this is OK?

I confirm that now the order of generated messages, is as desired and tests are still passing. The `VF` is back, so I also brought it back into the two tests.

I will look at the remaining issue and check whether `fix_nodepchk_flag` is needed. But that sounds optional, so I wanted to push what you have fixed, to unblock the PR.

@bryanpkc , the commit message has also been reworded.

**Comment by kiranchandramohan** [comment]
Thanks @michalpasztamobica, we need to test this well. Will also benefit from other reviewers having a closer look.

**Comment by michalpasztamobica** [comment]
@kiranchandramohan , I removed the body of the `fix_nodepchk_flag()` and that did not fail any test and did not influence the `.ll` output when building `ivdep.f90`. Any suggestions what else to check? Should I submit this as a separate patch for further verification with any available downstream tests?

**Comment by kiranchandramohan** [comment]
If the nodepcheck flow is not used anymore I would recommend creating a fix_access_group_flag function.

**Comment by michalpasztamobica** [comment]
Hi @kiranchandramohan ,
I am trying to figure out in detail how to approach the problem. Apologies, but I am still not very good with compiler-internals and the closer to LLVM, the bigger the gap in my knowledge.

Can you elaborate more on what the `fix_access_group_flag` would do? I assume the current `fix_nodepcheck_flang` function is allowing us to skip some particular dependency check between branches and that for some reason it seems not to make a difference any more now that access groups have been introduced. But we have switched from `llvm.mem.parallel_loop_accesses`, so I don't fully see the connection and why we even need to "fix" the `access_group` flag.

Sorry, if these are silly questions I am counting on your patience with me :)

File: tools/flang2/flang2exe/cgmain.cpp
Comment (michalpasztamobica) [reply tools/flang2/flang2exe/cgmain.cpp:2925]: This has been addressed by adding the flag in the appropraite section.

**Comment by kiranchandramohan** [comment]
I would like to understand what fix_nodepcheck_flag does before completely removing it from the flow. Alternatively, introduce a fix_access_group_flag function in place of fix_nodepcheck_flag.

**Review by Chirag-Khandelwal — COMMENTED** [review]
This looks good to me (quite clean too!). I'll let others take a look.
Verified the tests as well, and they run just fine for me :)

File: tools/flang2/flang2exe/cgmain.cpp
Comment (PeixinQiao) [inline tools/flang2/flang2exe/cgmain.cpp:1018]: Nit: Missing two spaces. Please use the following style.
```
    cached_vectorize_disable_metadata = ll_get_md_node(cpu_llvm_module,
        LL_PlainMDNode, lvcomp, 2);
```

File: tools/flang2/flang2exe/cgmain.cpp
Comment (PeixinQiao) [inline tools/flang2/flang2exe/cgmain.cpp:1033]: Nit: Missing two spaces. The same as above.

File: tools/flang2/flang2exe/cgmain.cpp
Comment (PeixinQiao) [inline tools/flang2/flang2exe/cgmain.cpp:1049]: Nit: Missing two spaces. The same as above.

File: tools/flang2/flang2exe/cgmain.cpp
Comment (PeixinQiao) [inline tools/flang2/flang2exe/cgmain.cpp:1382]: Nit: Missing two spaces. The same as above.

File: tools/flang2/flang2exe/cgmain.cpp
Comment (PeixinQiao) [inline tools/flang2/flang2exe/cgmain.cpp:1839]: Nit: Exceed 80 columns.

File: tools/flang2/flang2exe/cgmain.cpp
Comment (bryanpkc) [reply tools/flang2/flang2exe/cgmain.cpp:1018]: It's probably easier to run this hunk through clang-format, which would give:
```c
    cached_vectorize_disable_metadata =
        ll_get_md_node(cpu_llvm_module, LL_PlainMDNode, lvcomp, 2);
```

**Review by bryanpkc — APPROVED** [review]
LGTM, other than the formatting nits raised by @PeixinQiao. Our downstream build checked out fine too.

**Comment by michalpasztamobica** [comment]
Many thanks for all the reviews, I just pushed a commit addressing @PeixinQiao 's findings. I did not squash it, as it is kind of Peter' code originally. I hope this is fine?

@kiranchandramohan , seems like all reviews are in place and we could try to merge this in?

**Comment by bryanpkc** [comment]
@michalpasztamobica The formatting fixes should be squashed into the original commits that introduced those changes.

**Comment by michalpasztamobica** [comment]
@bryanpkc , alright, squashed :)

File: tools/flang2/flang2exe/cgmain.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/cgmain.cpp:1831]: Incorrect indentation. @michalpasztamobica  Could you also fix this one?

**Comment by kiranchandramohan** [comment]
> I would like to understand what fix_nodepcheck_flag does before completely removing it from the flow. Alternatively, introduce a fix_access_group_flag function in place of fix_nodepcheck_flag.

@bryanpkc what was your conclusion regarding this point?

File: tools/flang2/flang2exe/cgmain.cpp
Comment (michalpasztamobica) [reply tools/flang2/flang2exe/cgmain.cpp]: Done :)

**Comment by michalpasztamobica** [comment]
@bryanpkc , @kiranchandramohan , it looks like Kiran's question is the last blocker standing in the way of merging this in? Any update on this?

**Comment by bryanpkc** [comment]
As I mentioned at the last call, I tried to read the source and figure out what `fix_nodepcheck_flag` was doing, with not much success. I'll find time to dig some more and update this issue one way or another.

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

**Pull request merged**
